### PR TITLE
Fix worker dying on upload exception and shutdown

### PR DESCRIPTION
**Pull request by ihorkhavkin**
Fixes issue with upload stopping when `move_file_to_bucket` fails and cleans up shutdown a bit (as case doesn't seem to work on class comparison - http://stackoverflow.com/a/5694333/5657104 )

### CODE REVIEW

**Comment by ihorkhavkin** [comment]
https://github.com/logstash-plugins/logstash-output-s3/issues/79
Will sign CLA after my employer's approval.

**Comment by ihorkhavkin** [comment]
CLA was signed, please let me know if something else needs to be done for check ^ to pass

**Comment by andrewvc** [comment]
@ihorkhavkin thanks for the PR. Apologies for the extreme delay in looking at this, I'm taking a look now :)

**Comment by andrewvc** [comment]
This mostly looks good to merge to me, but I want to hold off for a minute. It looks to me like nothing properly terminates the worker threads on shutdown. That seems problematic. 

The tests all pass but I want to do a manual test to see how that goes. I'm not sure if this is the PR to fix this in (probably not) but I want to think about it over the weekend.

File: lib/logstash/outputs/s3.rb [hunk 1]
```diff
@@ -404,23 +404,36 @@ def configure_upload_workers
...
-        @upload_queue.enq(LogStash::ShutdownEvent)
+        @upload_queue.enq(LogStash::SHUTDOWN)
       else
         @logger.debug("S3: upload working is uploading a new file", :filename => File.basename(file))
         move_file_to_bucket(file)
+      end
+    rescue Exception => ex
+      @logger.error("failed to upload, will re-enqueue #{file} for upload",
+                    :ex => ex, :backtrace => ex.backtrace)
+      unless file.nil?
```
Comment (ph) [inline lib/logstash/outputs/s3.rb]: The file cannot be nil, since we are dequeuing stuff.

File: lib/logstash/outputs/s3.rb [hunk 2]
```diff
@@ -404,23 +404,36 @@ def configure_upload_workers
...
   private
   def upload_worker
-    file = @upload_queue.deq
+    file = nil
+    begin
+      file = @upload_queue.deq
 
-    case file
-      when LogStash::ShutdownEvent
+      if file.is_a? LogStash::ShutdownEvent
```
Comment (ph) [inline lib/logstash/outputs/s3.rb]: We should do the compare with the `LogStash::SHUTDOWN`

File: lib/logstash/outputs/s3.rb [hunk 3]
```diff
@@ -404,23 +404,36 @@ def configure_upload_workers
         while true do
           @logger.debug("S3: upload worker is waiting for a new file to upload.", :worker_id => worker_id)
 
-          upload_worker
+          begin
+            upload_worker
+          rescue Exception => ex
+            @logger.error('upload_worker unhandled exception', :ex => ex, :backtrace => ex.backtrace)
+            raise LogStash::Error, 'S3: uploader thread exited unexpectedly'
```
Comment (ph) [inline lib/logstash/outputs/s3.rb]: I don't think this line solve the problem of the thread dying when an exception is raised inside a thread.
I believe we should only catch the exception in the `#upload_worker` method and not in two different places.

File: lib/logstash/outputs/s3.rb (hunk 1 above)
Comment (ph) [inline lib/logstash/outputs/s3.rb]: Actually, It could be nil when queue `#deq` get interrupted at shutdown.

Comment (ph) [inline lib/logstash/outputs/s3.rb]: After discussing with @andrewvc we don't think this case happen in the JVM.

**Comment by andrewvc** [comment]
Merged in #90

**Pull request closed without merging**